test/src/unit-capi-smoke-test.cc

```diff
@@ -0,0 +1,757 @@
+      }
+      write_buffer_size = dim_buffer_len * sizeof(uint64_t);
+
+      write_query_buffers.emplace_back(
+          dims_iter->name_,
+          d_write_buffer,
+          &(write_buffer_size),
+          nullptr,
+          nullptr);
+      dims_iter = std::next(dims_iter, 1);
+    }
+
+    // Execute the write query.
+    write(array_name, write_query_buffers, write_order, encryption_type);
+
+    // Free the dimension write buffers vector
+    d_write_buffers.clear();
+    d_write_buffers.shrink_to_fit();
+  } else {
+    // Execute the write query.
+    write(array_name, write_query_buffers, write_order, encryption_type);
+  }
+
+  // Define the read query buffers for "a".
+  vector<test_query_buffer_t> read_query_buffers;
+
+  uint64_t a_read_buffer_size;
+  int32_t* a_read_buffer = nullptr;
+
+  if (test_attr.cell_val_num_ == TILEDB_VAR_NUM) {
+    a_read_buffer_size =
+        total_cells * 2 * tiledb_datatype_size(test_attr.type_);
+    a_read_buffer = (int32_t*)tdb_malloc(a_read_buffer_size);
+    for (uint64_t i = 0; i < total_cells * 2; i++) {
+      a_read_buffer[i] = 0;
+    }
+  } else {
+    a_read_buffer_size = total_cells * tiledb_datatype_size(test_attr.type_);
+    a_read_buffer = (int32_t*)tdb_malloc(a_read_buffer_size);
+    for (uint64_t i = 0; i < total_cells; i++) {
+      a_read_buffer[i] = 0;
+    }
+  }
+
+  uint64_t a_read_buffer_offset_size = total_cells * sizeof(uint64_t);
+  uint64_t* a_read_buffer_offset =
+      (uint64_t*)tdb_malloc(a_read_buffer_offset_size);
+  for (uint64_t i = 0; i < total_cells; i++) {
+    a_read_buffer_offset[i] = 0;
+  }
+
+  // Variable buffer is set only when the attribute is var-sized.
+  if (test_attr.cell_val_num_ == TILEDB_VAR_NUM) {
+    read_query_buffers.emplace_back(
+        test_attr.name_,
+        a_read_buffer_offset,
+        &a_read_buffer_offset_size,
+        a_read_buffer,
+        &a_read_buffer_size);
+  } else {
+    read_query_buffers.emplace_back(
+        test_attr.name_, a_read_buffer, &a_read_buffer_size, nullptr, nullptr);
+  }
+
+  uint64_t subarray_len = 2 * test_dims.size();
+  uint64_t subarray_size = subarray_len * sizeof(uint64_t);
+  uint64_t* subarray_full = (uint64_t*)tdb_malloc(subarray_size);
+  for (uint64_t i = 0; i < subarray_len; i += 2) {
+    const uint64_t min_range = ((uint64_t*)(test_dims[i / 2].domain_))[0];
+    const uint64_t max_range = ((uint64_t*)(test_dims[i / 2].domain_))[1];
+    subarray_full[i] = min_range;
+    subarray_full[i + 1] = max_range;
+  }
+
+  read(array_name, read_query_buffers, subarray_full, encryption_type);
+
+  // Ensure that each value in `a_read_buffer` corresponds to its index in
+  // the original `a_write_buffer`.
+  REQUIRE(!memcmp(a_read_buffer, a_write_buffer, a_read_buffer_size));
+
+  // Free the write buffers vector
+  write_query_buffers.clear();
+  write_query_buffers.shrink_to_fit();
+
+  // Free the read buffers vector
+  read_query_buffers.clear();
+  read_query_buffers.shrink_to_fit();
```

Comment:
it isn't necessary to clear these vectors, however it is necessary to free all of the malloc'd buffers they contain